File: solutions/2391. Minimum Amount of Time to Collect Garbage/2391.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  int garbageCollection(
    const vector<string>& garbage, 
    const vector<int>& travel
  ) {
    vector<int> prefix(travel.size());
    partial_sum(travel.begin(), travel.end(), prefix.begin());
    const int timeM = getTime(garbage, prefix, 'M');
    const int timeP = getTime(garbage, prefix, 'P');
    const int timeG = getTime(garbage, prefix, 'G');

    return timeM + timeP + timeG;
  }

 private:
  int getTime(
    const vector<string>& garbage, 
    const vector<int>& prefix, 
    char c
  ) {
    int characterCount = 0;
    int lastIndex = -1;

    for (int i = 0; i < garbage.size(); ++i) {
      string_view s = garbage[i];
      if (ranges::any_of(s, [c](const char g) { return g == c; }))
        lastIndex = i;
      characterCount += ranges::count(s, c);
    }

    return characterCount + (lastIndex <= 0 ? 0 : prefix[lastIndex - 1]);
  }
};

File: solutions/2391. Minimum Amount of Time to Collect Garbage/2391.rs
// Time complexity: O(n)
// Space complexity: O(n)
impl Solution {
    pub fn garbage_collection(garbage: Vec<String>, travel: Vec<i32>) -> i32 {
        let prefix: Vec<i32> = travel
            .iter()
            .scan(0, |acc, &x| {
                *acc += x;
                Some(*acc)
            })
            .collect();

        let get_time = |c: char| -> i32 {
            let mut character_count = 0;
            let mut last_index = None;

            for (i, s) in garbage.iter().enumerate() {
                if s.contains(c) {
                    last_index = Some(i);
                }
                character_count += s.chars().filter(|&ch| ch == c).count() as i32;
            }

            character_count + last_index.map_or(0, |i| if i == 0 { 0 } else { prefix[i - 1] })
        };

        get_time('M') + get_time('P') + get_time('G')
    }
}
